### ISSUE TITLE
[EKS+tunnel] Wrong source address of return packet in host-to-pod

### ISSUE THREAD

**Issue by ti-mo** [cilium/cilium#16975]
## Bug report

**General Information**

I'm running DNS queries from the Cilium agent pod (192.168.146.198) against CoreDNS running on a remote node. (10.0.1.120) `nslookup` times out.

```
13:42:43.672847 IP 192.168.146.198.49128 > 10.0.1.120.53: 11112+ AAAA? jenkins.cilium.io. (35)
13:42:43.673706 IP 192.168.186.218.53 > 192.168.146.198.49128: 11112* 0/1/0 (112)
```

The reply packet is incorrectly SNATed to the source IP of the other node (192.168.186.218), breaking the connection.

Note that this breaks the FQDN proxy in tunnel mode because DNS lookups are performed from the agent pod.

- Cilium version: 1.10.3
- Kernel version: 5.4.129-62.227.amzn2.x86_64
- Sysdump attached
[cilium-sysdump-20210722-164747.zip](https://github.com/cilium/cilium/files/6863101/cilium-sysdump-20210722-164747.zip)

**How to reproduce the issue**

1. Set up EKS cluster
2. `./cilium install --datapath-mode=tunnel --ipam cluster-pool`
3. Drop into Cilium agent pod and nslookup against a CoreDNS instance on another node.

Possibly related:
- https://github.com/cilium/cilium/issues/16963
- https://github.com/cilium/cilium/issues/16938

**After fixing the issue**

- Test exceptions added to EKS (Tunnel) workflow in `cilium-cli` by https://github.com/cilium/cilium-cli/pull/450 need to be removed.

**Comment by nbusseneau** [comment]
Adding mention in OP to update `cilium-cli` when fixed.

**Comment by pchaigno** [comment]
**TL;DR.** I've deployed a cluster with the same Cilium version, using the same commands, and I'm unable to reproduce. I can also successfully run the connectivity tests.

### Reproduction Steps

```
$ kslogs | grep 1.10.3
level=info msg="Cilium 1.10.3 4145278 2021-07-15T16:11:03+02:00 go version go1.16.6 linux/amd64" subsys=daemon
level=info msg="Cilium 1.10.3 4145278 2021-07-15T16:11:03+02:00 go version go1.16.6 linux/amd64" subsys=daemon
$ kslogs | grep tunnel=
level=info msg="  --tunnel='vxlan'" subsys=daemon
level=info msg="  --tunnel='vxlan'" subsys=daemon

$ ks get pods -o wide -l k8s-app=kube-dns
NAME                      READY   STATUS    RESTARTS   AGE    IP           NODE                                           NOMINATED NODE   READINESS GATES
coredns-996495cbb-rqbkb   1/1     Running   0          179m   10.0.0.155   ip-192-168-132-89.eu-west-2.compute.internal   <none>           <none>
coredns-996495cbb-xxqj4   1/1     Running   0          179m   10.0.0.48    ip-192-168-132-89.eu-west-2.compute.internal   <none>           <none>
$ ks exec cilium-zq5zf -- ip a show dev eth0 | grep "inet "
    inet 192.168.184.76/19 brd 192.168.191.255 scope global dynamic eth0
$ ks exec cilium-zq5zf -- dig @10.0.0.155 +short jenkins.cilium.io
54.148.123.155
```

### Investigation
I also traced the packets while sending the request and all seems to work fine:
```
$ ks exec cilium-zq5zf -- cilium monitor | grep :53
level=info msg="Initializing dissection cache..." subsys=monitor
<- host flow 0x0 identity host->unknown state new ifindex 0 orig-ip 0.0.0.0: 10.0.1.69:42187 -> 10.0.0.155:53 udp
-> overlay flow 0x0 identity remote-node->unknown state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.0.1.69:42187 -> 10.0.0.155:53 udp
<- overlay flow 0x5a055a05 identity unknown->unknown state new ifindex cilium_vxlan orig-ip 0.0.0.0: 10.0.0.155:53 -> 10.0.1.69:42187 udp
-> stack flow 0x5a055a05 identity unknown->unknown state new ifindex cilium_host orig-ip 0.0.0.0: 10.0.0.155:53 -> 10.0.1.69:42187 udp
```
I also tried the same using the CoreDNS VIP instead of the backend pod IP address.

The packet path is as expected in this configuration:
1. The request from the hostns is assigned source IP 10.0.1.69 and sent to `cilium_host`:
```
$ ks exec cilium-zq5zf -- ip r get 10.0.0.155 
10.0.0.155 dev cilium_host src 10.0.1.69 uid 0 
    cache mtu 8951
```
2. `bpf_host` attached to `cilium_host` then sends the packet over the tunnel.
3. On the destination node, the packet is delivered to the lxc device.
4. The reply has destination IP address 10.0.1.69, according to the destination node's ipcache, it is sent to tunnel endpoint 192.168.184.76 (the source node) using a redirect to the overlay device:
```
$ ks exec cilium-p9q8b -- cilium bpf ipcache list | grep 10.0.1.69
10.0.1.69/32        6 0 192.168.184.76
```
I checked that all of the above state is the same in the provided sysdump (with different IP addresses of course), so the same should have happened.

### Next Steps
Maybe I missed something and it will be clear from reading the above. A few things are not explicitly stated in the original post:
- What was the exact `nslookup` command used?
- Which version of the Cilium CLI was used?
- On which interface was the tcpdump output collected?

If none of the above helps spot the difference, then next step is probably to revert https://github.com/cilium/cilium-cli/pull/450.
@nbusseneau Did we have any sysdump for that test failure? That could also provide another clue.

**Comment by nbusseneau** [comment]
Sysdumps from GH workflow runs have expired :/
Looking at history / timestamps though, CLI version was v0.8.5.

**Comment by pchaigno** [comment]
Thanks @nbusseneau! I've tested with cilium-cli v0.8.5 and I'm getting the same result.

**Comment by joamaki** [comment]
I think I'm hitting this issue in my PR (https://github.com/cilium/cilium-cli/pull/1015) adding cilium-health check to the connectivity test. "cilium-health status --probe" reports OK for local health endpoint, but fails to connect to remote endpoint:
```
  cilium-cilium-cli-2825580078-tunnel/ip-192-168-106-152.us-east-2.compute.internal:
    Host connectivity to 192.168.106.152:
      ICMP to stack:   OK, RTT=800.479µs
      HTTP to agent:   OK, RTT=1.073148ms
    Endpoint connectivity to 10.0.1.2:
      ICMP to stack:   Connection timed out
      HTTP to agent:   Get "http://10.0.1.2:4240/hello": dial tcp 10.0.1.2:4240: i/o timeout (Client.Timeout exceeded while awaiting headers)
```

**Comment by pchaigno** [comment]
@joamaki What makes you think it's the same issue? Do you have reproduction steps (see above, I tried to reproduce before without success)?
